music21/musicxml/m21ToXml.py

```diff
@@ -3265,55 +3280,88 @@ def parseFlatElements(
+        # Group all objects by offsets and then do a different order than normal sort.
+        # That way chord symbols and other 0-width objects appear before notes as much as
         # possible.
-        objGroup: list[base.Music21Object]
         objIterator: OffsetIterator[base.Music21Object] = OffsetIterator(m)
         for objGroup in objIterator:
             groupOffset = m.elementOffset(objGroup[0])
-            amountToMoveForward = int(round(divisions * (groupOffset
-                                                             - self.offsetInMeasure)))
-            if amountToMoveForward > 0 and any(
+            offsetToMoveForward = groupOffset - self.offsetInMeasure
+            if offsetToMoveForward > 0 and any(
                     isinstance(obj, (note.GeneralNote, clef.Clef)) for obj in objGroup):
                 # gap in stream between GeneralNote/Clef objects: create <forward>
-                mxForward = Element('forward')
-                mxDuration = SubElement(mxForward, 'duration')
-                mxDuration.text = str(amountToMoveForward)
-                root.append(mxForward)
-                self.offsetInMeasure = groupOffset
+                self.moveForward(offsetToMoveForward)
 
             notesForLater = []
             for obj in objGroup:
                 # we do all non-note elements (including ChordSymbols not written as chord)
                 # first before note elements, in musicxml
+                if isinstance(obj, spanner.SpannerAnchor):
+                    # nothing to do with this (only prePostObjectSpanners, and
+                    # we will do that below)
+                    continue
+
                 if isinstance(obj, note.GeneralNote) and (
                     not (isHarm := isinstance(obj, harmony.Harmony))
                     or (isHarm and obj.writeAsChord)
                 ):
                     notesForLater.append(obj)
                 else:
-                    self.parseOneElement(obj)
+                    self.parseOneElement(obj, skipPrePostObjectSpanners=hasSpannerAnchors)
 
             for n in notesForLater:
                 if n.isRest and n.style.hideObjectOnPrint and n.duration.type == 'inexpressible':
                     # Prefer a gap in stream, to be filled with a <forward> tag by
                     # fill_gap_with_forward_tag() rather than raising exceptions
                     continue
-                self.parseOneElement(n)
+                self.parseOneElement(n, skipPrePostObjectSpanners=hasSpannerAnchors)
+
+        # Second pass: If this flat stream (measure/voice) contains any SpannerAnchors, we
+        # will perform a second pass, emitting any pre/postList for any object that is
+        # functioning as a spanner anchor (the SpannerAnchors, as well as any GeneralNotes
+        # that start or end spanners).
+
+        if hasSpannerAnchors:
+            # return to the beginning of the measure, to emit any SpannerAnchors.
```

Comment:
at the end of this, should we jump forward to where we were before backing up?  In case there is more to do -- because backupAfterwards setting might suppose that we're at the end of the measure.

Reply:
I'm not sure.  In the emitted MusicXML we _are_ at the end of the measure (although, self.offsetInMeasure might be at an offset before the furthest offset in the measure, if the spanners didn't reach the end of the measure).  Does that matter?  Is it worth adding yet another `<forward>` tag to the emitted MusicXML for that?  Hmmm... looking at mainElementsParse, it looks like the only thing that might end up at the wrong offset is the rightBarline.  Do you need a `<forward>` before a barline to get the barline to be at the right offset?

Reply:
But that rightBarline error might also happen if a second voice is shorter than a first voice.  So we're not making anything worse????

Reply:
I'm currently thinking we should jump forward only if (1) the spanner pass ended up short of the notes pass AND (2) backupAfterwards is False.  We should never jump backward (even if it could happen that the spanner pass was longer duration than the notes pass, which I think it couldn't).

Reply:
done.